### PR TITLE
feat(framework): Store traffic in `LinkState`

### PR DESCRIPTION
**Pull request by chongshenng**
Merge after:
* #6244

### CODE REVIEW

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py [hunk 1]
```diff
@@ -778,3 +778,39 @@ def store_traffic(
...
+                run_id,
+                bytes_sent,
+                bytes_recv,
+            )
+            return
+
+        # Skip if no traffic to record
+        if bytes_sent == 0 and bytes_recv == 0:
+            log(WARNING, "No traffic data to store for run_id %d", run_id)
+            return
```
Comment (jafermarq) [inline framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: when would this be the case?

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py [hunk 2]
```diff
@@ -778,3 +778,39 @@ def store_traffic(
...
+        # Validate non-negative values
+        if bytes_sent < 0 or bytes_recv < 0:
+            log(
+                WARNING,
+                "Negative traffic values for run %d: bytes_sent=%d, bytes_recv=%d",
+                run_id,
+                bytes_sent,
+                bytes_recv,
+            )
+            return
```
Comment (jafermarq) [inline framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: should this be treated as an error? the function that would call this will not know an issue (negative bytes) happened.

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py (hunk 1 above)
Comment (chongshenng) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: This will be a very edge case, if the `ServerApp` and `ClientApp` is not sending anything, e.g. if some computation failed in the `ClientApp` and the user opts to return nothing in the `Message`. IMO, it's necessary to log this.

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py (hunk 2 above)
Comment (jafermarq) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: I think logging as error is good but what i meant is to actually treat is as an error and raise an exception that can be handled in the function that calls this one. Without raising an excp. the caller will have no idea something was wrong (since the return type is always `None`)

Comment (jafermarq) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: likewise for the other scenarios where `None` is returned after logging an `ERROR`

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py (hunk 1 above)
Comment (jafermarq) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: should we raise an exception in addition to log an error? else the caller doesn't know somethign is very wrong (negative bytes)

Comment (chongshenng) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: @jafermarq I think your comment was meant for [the other thread](https://github.com/adap/flower/pull/6243#discussion_r2592851272) where the code originally logged an `ERROR`. IMO it's better to raise an exception since it already contains the error message. We can handle the error in `FleetServicer` when `state.store_traffic()` is called. Wdyt?

Comment (chongshenng) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: Updated to raise `ValueError` for negative values and when `run_id` is not found.

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py (hunk 2 above)
Comment (chongshenng) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py]: Updated to raise `ValueError` for negative values and when `run_id` is not found.

File: framework/py/flwr/server/superlink/linkstate/linkstate_test.py
Comment (jafermarq) [inline framework/py/flwr/server/superlink/linkstate/linkstate_test.py:1]: for completeness. should we include a test that attempts to call `.store_traffic()` to a run that doesn't exist?

Comment (danieljanes) [reply framework/py/flwr/server/superlink/linkstate/linkstate_test.py:1]: We also need a test that prevents negative values from being added

Comment (chongshenng) [reply framework/py/flwr/server/superlink/linkstate/linkstate_test.py:1]: Added

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py [hunk 3]
```diff
@@ -776,3 +776,24 @@ def get_serverapp_log(
 
     def store_traffic(self, run_id: int, *, bytes_sent: int, bytes_recv: int) -> None:
```
Comment (panh99) [inline framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:777]: Quick q: when calling `LinkState.store_traffic`, will we ever provide non-zero values for both `bytes_sent` and `bytes_recev`? If not, should we set default values for the two arguments?

File: framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py [hunk 4]
```diff
@@ -1263,6 +1263,40 @@ def get_valid_message_ins(self, message_id: str) -> dict[str, Any] | None:
...
+            if not rows:
+                raise ValueError(f"Run {run_id} not found")
+
+            # Perform the update
+            update_query = """
+                    UPDATE run
+                    SET bytes_sent = bytes_sent + ?,
+                        bytes_recv = bytes_recv + ?
+                    WHERE run_id = ?;
+                """
```
Comment (panh99) [inline framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py]:
```suggestion
                    UPDATE run
                    SET bytes_sent = bytes_sent + ?,
                        bytes_recv = bytes_recv + ?
                    WHERE run_id = ?
                    RETURNING run_id;
                """
```

File: framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py [hunk 5]
```diff
@@ -1263,6 +1263,40 @@ def get_valid_message_ins(self, message_id: str) -> dict[str, Any] | None:
...
+            # Check if run exists
+            query = """
+                    SELECT run_id
+                    FROM run
+                    WHERE run_id = ?;
+                """
+            rows = self.conn.execute(query, (sint64_run_id,)).fetchall()
+
+            if not rows:
+                raise ValueError(f"Run {run_id} not found")
```
Comment (panh99) [inline framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py]: It would be good to reduce the number of queries when possible. We can use `RETURNING` keyword to check if the row exist. See the change below.

File: framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py (hunk 4 above)
Comment (panh99) [reply framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py]: And then we can check if returned rows is not empty to see if the run row exists.

File: framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py (hunk 3 above)
Comment (chongshenng) [reply framework/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:777]: No, we don't ever set both to zero at the same time. @danieljanes brought up a point about not setting default values here as it could be introduce subtle bugs, which I tend to agree. So, the acceptable usage here would be
* `LinkState.store_traffic(run_id: ..., bytes_sent=<val>, bytes_recv=0)`
* `LinkState.store_traffic(run_id: ..., bytes_sent=0, bytes_recv=<val>)`

File: framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py [hunk 6]
```diff
@@ -1263,6 +1263,35 @@ def get_valid_message_ins(self, message_id: str) -> dict[str, Any] | None:
...
+        sint64_run_id = uint64_to_int64(run_id)
+
+        with self.conn:
+            # Check if run exists, performing the update only if it does
+            update_query = """
+            UPDATE run
+            SET bytes_sent = bytes_sent + ?,
+                bytes_recv = bytes_recv + ?
+            WHERE run_id = ?
+            RETURNING run_id;
```
Comment (panh99) [inline framework/py/flwr/server/superlink/linkstate/sqlite_linkstate.py]:
```suggestion
                UPDATE run
                SET bytes_sent = bytes_sent + ?,
                    bytes_recv = bytes_recv + ?
                WHERE run_id = ?
                RETURNING run_id;
```
